src/gridfs.rs

```diff
@@ -0,0 +1,168 @@
+use core::task::{Context, Poll};
+use std::{
+    io::{self, Result},
+    marker::PhantomPinned,
+    pin::Pin,
+};
+
+use crate::{
+    concern::{ReadConcern, WriteConcern},
+    cursor::Cursor,
+    selection_criteria::ReadPreference,
+    Database,
+};
+
+use bson::{oid::ObjectId, DateTime, Document};
+use serde::Deserialize;
+use typed_builder::TypedBuilder;
+
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+#[derive(Clone, Debug, Default, Deserialize, TypedBuilder)]
+#[builder(field_defaults(setter(into)))]
+#[non_exhaustive]
+pub struct GridFSBucketOptions {
+    bucket_name: Option<String>,
+    chunk_size_bytes: Option<i32>,
+    write_concern: Option<WriteConcern>,
+    read_concern: Option<ReadConcern>,
+    read_preference: Option<ReadPreference>,
+}
+
+pub struct GridFSUploadOptions {
+    chunk_size_bytes: Option<i32>,
+    metadata: Option<Document>,
+}
+
+pub struct GridFSDownloadByNameOptions {
+    revision: Option<i32>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, TypedBuilder)]
+#[builder(field_defaults(setter(into)))]
+#[non_exhaustive]
+pub struct GridFSFindOptions {
+    allow_disk_use: Option<bool>,
+    batch_size: Option<i32>,
+    limit: Option<i32>,
+    max_time_ms: Option<i64>,
+    no_cursor_timeout: Option<bool>,
+    skip: i32,
+    sort: Option<Document>,
+}
+
+// Contained in a "chunks" collection for each user file
+pub struct Chunk<T: Eq + Copy> {
+    id: ObjectId,
+    files_id: T,
+    n: i32,
+    // default size is 255 KiB
+    data: Vec<u8>,
+}
+
+// A collection in which information about stored files is stored. There will be one files
+// collection document per stored file.
+pub struct FilesCollectionDocument<T: Eq + PartialEq + Copy> {
+    id: T,
+    length: i64,
+    chunk_size: i32,
+    upload_date: DateTime,
+    filename: String,
+    metadata: Document,
+}
+
+pub struct GridFSBucket {
+    // Contains a "chunks" collection
+    pub db: Database,
```

Comment:
As per the last line in [this section](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#configurable-gridfsbucket-class) of the spec, bucket instances should be immutable. For that reason we'll want to make these fields non-`pub` so that users can't declare a `mut` bucket and change them. We may want to add getter methods for the read concern/write concern/selection criteria (similar to those of [`Collection`](https://docs.rs/mongodb/latest/mongodb/struct.Collection.html#)) so that users can see what those were resolved to.